deform/tests/test_widget.py

```diff
@@ -2339,6 +2339,16 @@ def test_optgroup_and_tuple(self):
         self.assertEqual(normalized[1].label, "label")
         self.assertEqual(normalized[1].options, (("2", "two"),))
 
+    def test_generator_choices(self):
+        def choices_generator():
+            for value in range(2):
+                yield (value, value)
+
+        choices = choices_generator()
+        normalized = self._call(choices)
+        self.assertEqual(normalized[0], ("0", 0))
+        self.assertEqual(normalized[1], ("1", 1))
```

Comment:
Bug will present itself when you run:

```
         normalized = self._call(choices)
         self.assertEqual(normalized[0], ("0", 0))
         self.assertEqual(normalized[1], ("1", 1))
```

After this again.

Where you've placed the `list()` won't help, the issue is also not in `normalize_choices`. Instead the select widget should call `list()` on `values` or the docs should be updated to not mention sequence but instead clarify it needs to be a list.

Reply:
I tried calling `list()` on the values, but the issue persisted because all of the generator's values were consumed on the first pass. It looks like fixing docstrings is part of the fix.

I tried changing it in commit 4494b33, and that seems to be OK. I'd appreciate your feedback.

I really wanted to test the error message, too, but I could not figure it out. I tried a context manager per `assertRaises`'s docstrings, but that didn't work.

Reply:
w00t! I fingered it out. Context managers are tricksy. I got the idea from https://stackoverflow.com/a/62253324/2214933

I think I need to apply this to radios and checkboxes as well, and update their docstrings, and then it should be good to go.

Reply:
Also I think it is OK to use a sequence type (list, tuple, or range), a set, or a dict for values. All appear to work, but one should probably stick to a sequence to retain the order of values.

Reply:
@bertjwregeer I'd appreciate your review on the final version.